Ruff rules update

martinhoyer opened the pull request:
Adding a few more linter checks to pre-commit/ruff.
There are more useful rules to implement, omitting for sake of not having too many changes in one PR.

### [flake8-bandit](https://docs.astral.sh/ruff/rules/#flake8-bandit-s)
Using `tempfile.mkdtemp` to securely create temporary directories. When `dir` argument of [mkdtemp](https://docs.python.org/3/library/tempfile.html#tempfile.mkdtemp) is not set: 

> "The default directory is chosen from a platform-dependent list, but the user of the application can control the directory location by setting the TMPDIR, TEMP or TMP environment variables."

This, instead of hardcoding `/tmp` or `/var/tmp` should make things also more platform-agnostic I presume.
As for `/var/tmp/tmt` - is something blocking making it also dynamic? Reminding that rpm is also not happy about it's usage: #2803 

Enabling jinja2 autoescape.
Ignoring unsecure xml parsing. (?)

### [flake8-builtins](https://docs.astral.sh/ruff/rules/#flake8-builtins-a)
Renamed some variables and ignored others. Probably can be improved - suggestions welcome.
Perhaps, where applicable, we could use trailing underscore? `shadowing_var` -> `shadowing_var_`?

### [flake8-implicit-str-concat](https://docs.astral.sh/ruff/rules/#flake8-implicit-str-concat-isc)
As recommended by PEP 8

### [flake8-logging](https://docs.astral.sh/ruff/rules/#flake8-logging-log) and [flake8-logging-format](https://docs.astral.sh/ruff/rules/#flake8-logging-format-g)
`warn` -> `warning`.
Allowing [f-strings in logging](https://docs.astral.sh/ruff/rules/logging-f-string/), as it's being used in many places and is not trivial to 'fix'.

### [PYI](https://docs.astral.sh/ruff/rules/#flake8-pyi-pyi), [ICN](https://docs.astral.sh/ruff/rules/#flake8-import-conventions-icn), [TID](https://docs.astral.sh/ruff/rules/#flake8-tidy-imports-tid), [INT](https://docs.astral.sh/ruff/rules/#flake8-gettext-int), [Q003, Q004](https://docs.astral.sh/ruff/rules/#flake8-quotes-q) and [PLR](https://docs.astral.sh/ruff/rules/#refactor-r)

I suspect the `else: if:` -> `elif` to reduce indentation can be controversial, happy to revert.

LecrisUT commented:
> Enabling jinja2 autoescape.

Careful with this one not to pick up the html version of autoescape

lukaszachy commented on tmt/steps/execute/internal.py:
```diff
@@ -32,7 +32,7 @@
 
 #: The default directory for storing test pid file.
-TEST_PIDFILE_ROOT = Path('/var/tmp')
+TEST_PIDFILE_ROOT = '/var/tmp'  # noqa: S108
```
What is wrong with making it Path here?

martinhoyer replied:
Nothing wrong. It just made more sense to me when I was doing the changes. Can revert.

happz replied:
Please, do so, it's not just any string, it's a *path* :)

martinhoyer replied:
Will do.
And it needs to be /var/tmp, yes?
I assume the whole 'needs to survive reboot' is not super importand for pidfile dir. 
`mkdtemp` with a prefix perhaps? :)

LecrisUT replied:
Speaking of which, shouldn't it be user-specific. What if this is run in an hpc with multiple simultaneous users? (Would have more issues with installation and stuff, but still)

happz replied:
It must be possible to find this directory not just by tmt, but also by tools that are running outside of the context of a test when an out-of-session reboot is needed. `tmt-reboot` executed by a user in their SSH session must be able to find a directory that tmt used to store pid of the test it's running in completely different SSH session. Therefore it seems to be the predictability is necessary - I suppose could get away with something like `/var/run` and so on, there might be (solvable) permission issues, but it needs to be predictable, i.e. I don't see `mkdtemp` working reliably.

happz replied:
It may become user-specific, but then again, let's say guests A, B and C are running their tests on a guest, and user A runs `tmt-reboot` in another of their SSH session. Per-user pid file would be fine, it'd nicely notify A's tmt to stop the test and reboot the guest - but B's and C's tmt runs would be left in the dark. I suppose this fits the shelf of issues related to teaching tmt to share a guest with other unrelated tmts, as there are no strong guarantees when it comes to sharing...

martinhoyer replied:
Path is back; Keeping /var/tmp/.

lukaszachy reviewed (commented):
As for the /var/tmp/tmt usage: do you have any other location which:
Survives reboot, can hold a lot of data, is r/w by everyone?

The current insecure state is by design, it could be improved for sure but we need to support:
- multi user (more users running tmt on the same machine). In the past the first user running tmt has created the run directory structure and locked out others.
- user mapping between guest and host

martinhoyer commented:
> As for the /var/tmp/tmt usage: do you have any other location which: Survives reboot, can hold a lot of data, is r/w by everyone?
> 
> The current insecure state is by design, it could be improved for sure but we need to support:
> 
>     * multi user (more users running tmt on the same machine). In the past the first user running tmt has created the run directory structure and locked out others.
> 
>     * user mapping between guest and host

Honestly I'm not an expert on unix dir conventions, but my first instinct is to say `/opt` ?

martinhoyer commented:
> > Enabling jinja2 autoescape.
> 
> Careful with this one not to pick up the html version of autoescape

Thanks @LecrisUT, do you have any suggestions on how to improve it please?

martinhoyer commented:
> Honestly I'm not an expert on unix dir conventions, but my first instinct is to say `/opt` ?

...which would be wrong.
Looking at https://refspecs.linuxfoundation.org/FHS_3.0/fhs/ch05.html, I'm guessing `/var/cache/tmt`

LecrisUT commented:
> I'm guessing `/var/cache/tmt`

I would vote for keeping it to `/var/tmp/tmt` and silencing the checks with `rpmlintrc`. `tmt` temp files seem quite complex: persistent after reboot, caching, writable by all users, etc. There is also already `TMT_WORKDIR_ROOT` to overwrite this, and hopefully the [system/user settings](https://tmt.readthedocs.io/en/stable/stories/cli.html#config) whenever that lands (would be good for stuff like `homebrew` packaging).

The other option would be using `/var/lib/tmt/...` or separating the locations into more standard locations like `/var/cache` etc., but I think that is less intuitive for finding and pruning temp files, not to mention even less portable.

martinhoyer commented on tmt/utils.py:
```diff
@@ -6560,7 +6561,10 @@ def default_template_environment() -> jinja2.Environment:
-    environment = jinja2.Environment()
+    # S701: `autoescape=False` is dangerous and can lead to XSS.
+    environment = jinja2.Environment(
+        autoescape=jinja2.select_autoescape(enabled_extensions='html', default_for_string=False)
```
Here it's not being used for xml, no?  
Not sure about explicit `default=False` tbh. Do you think it's realistic they would change it in the future?

LecrisUT replied:
> Here it's not being used for xml, no?

Probably not right now, but the escaping mechanism is similar enough for `html` and `xml` I think it's fine to include it rather than being surprised later on. It may be used for `junit` and other reporting mechanism later on, maybe even for plugins.

> Not sure about explicit `default=False` tbh. Do you think it's realistic they would change it in the future?

Jinja changing it like that, realistically not, it would be quite a dick move. But it's also a bit documentative that we really do not want to autoescape in general, particularly on how it impacts the general usecase of the template in the context of `fmf` and `yaml` files.

For the record if you want to reproduce how this affects, here is one yaml file that bit me when I've tried to add that:
```yaml
vars+:
  Extensions:
    - extras: " \" 1E-12\" "
```
Expanded in a Jinja template that basically used `{{ extras }}`. There the `\"` quote marks got changed to html escaped values.

happz replied:
If it's necessary to not apply autoescaping just to some templates, then be aware that `render_template()` and `default_template_environment()` do not know what they will be rendering, and they may render various outputs: `report/html` uses them to render actual HTML content, and `export -h rest` to render a ReST representation of a story. Then there's the `export -h template` with a fully arbitrary template controlled by a user, it may be a plain-text file or XML, we don't know. Plus, often the environment is not used for loading the template itself, i.e. no chance for it to see the file extension (`return render_template(template_filepath.read_text(), ...). Enabling it here probably won't have the effect you might expect, and may have side effects.

happz replied:
I would maybe turn it off, note it somewhere and make sure users would be aware of it... Or is there a way to make it enabled & correct for all possible templates? We already use it to render asciidoc and rest as well as HTML and neither of those two functions is aware of the nature of those templates...

LecrisUT replied:
With the `enabled_extensions=['html', 'xml']` it should only use it on `.html`/`.xml` output files. Definitely needs to be tested though. I do see value in having it enabled for things like `report/html`

happz replied:
Sure, I just point out that the environment knows nothing about the context. See https://github.com/teemtee/tmt/blob/main/tmt/steps/report/html.py#L100, this is how the template is rendered and saved. Environment is not aware of the destination file or its suffix.

I don’t dispute the value of autoescaping, just the way it would be enabled - there are no files the environment could inspect and decide based on their suffixes. `report/html` needs to tell to `render_template()` & co. “Guys, we’re rendering HTML, enable autoescaping, chop chop”. I suppose it would become a new parameter for the rendering family, would that make sense?

LecrisUT replied:
> autoescape
>
>    If set to True the XML/HTML autoescaping feature is enabled by default. For more details about autoescaping see Markup. As of Jinja 2.4 this can also be a callable that is passed the template name and has to return True or False depending on autoescape should be enabled by default.

My reading of this is that:
- `template_name` is a string -> use `default_for_string`
- `template_name` has extension -> check `enabled/disabled_extensions` -> fallback to `default`

martinhoyer replied:
@happz Not sure about keeping it off, based on what's being written in the Jinja API doc: "However autoescaping is not yet enabled by default though this will most likely change in the future. It’s recommended to configure a sensible default for autoescaping."
wdyt?

@LecrisUT I remembered why I only put html there - if we add 'xml' (or leave default, which is `['html', 'htm', 'xml']`), generating the docs fails.

LecrisUT replied:
> @LecrisUT I remembered why I only put html there - if we add 'xml' (or leave default, which is `['html', 'htm', 'xml']`), generating the docs fails.

Actually that is great news. we can see exactly how this affects and if it's `template_name` vs `output_file`, etc. The `xml` part is unexpected for the site generation and would be interesting to hunt down what is going there.

martinhoyer replied:
fwiw, it fails on url link: `tmt/docs/spec/hardware.rst:167: ERROR: Unknown target name: "pci id repository &lt;https://pci-ids.ucw.cz/&gt;".`

martinhoyer replied:
> it seems to me you see some connection
> Can you explain that to me like I'm five?

@happz Don't know about others, but I thought it was implied clearly that I have no idea what's going on. 
I'm going through the templates code now and slowly starting to understand, but still - you telling me how to deal with the flake8 error is all I can hope for. Just trying to help.

The jinja doc imho makes it clear not using autoescape is bad (as well as the flake8-bandit), but I do not know how to deal with things like "from url", so if you say autoescape=False is the best way to go, at least for now all things considered, I'm happy :)

LecrisUT replied:
Actually I think there is an issue with the implementation, but I need to look more closely. A few notes:
- `loader=FileSystemLoader()` should be used
  Jinja supports splitting a template into multiple files, and this loader is needed to be able to read those templates
- `env.get_tempalte(Path())`
  This should be coupled with the usage of `loader` above
- Looking with one implementation I have [here](https://github.com/LecrisUT/fmf-jinja/blob/66d243830871f8a5eecab718ae53f69e23f1d68c/src/fmf_jinja/template.py#L91-L96) (`autoescape` is wrong over there), indeed nowhere does it have access to the output file, only the template
- Regarding `.html.j2` vs `.j2.html`, if upstream doesn't support it, we should be able to write the callable to support that from their implementation.

> fwiw, it fails on url link: `tmt/docs/spec/hardware.rst:167: ERROR: Unknown target name: "pci id repository &lt;https://pci-ids.ucw.cz/&gt;".`

Is that with `default_for_string=False`? That is when running `sphinx` or which command? Should be able to traceback that and see in the debugger what produced that.

LecrisUT replied:
> I for one would leave such a task for a standalone patch. It won't be an oneliner and it would go well beyond a mere addition of ruff rules. Worth fixing though, unless you guys would work it, I would, to get the escaping right eventually.

Agreed let's revert and comment out (`S701`?) as a planned improvement here. I've opened an issue to track it: https://github.com/teemtee/tmt/issues/2873

martinhoyer replied:
@happz Feel free to rewrite the comment if needed :)

martinhoyer commented:
Updated, rebased.  
@happz Could you please take a look at the https://github.com/teemtee/tmt/pull/2842/commits/f836acbd063a2d62f6bed9f781be37cba7c70870 ?
The [docs](https://docs.astral.sh/ruff/rules/quoted-annotation/) mention this rule applies to \_\_future__ annotations usage, but I presume we see it due to https://github.com/astral-sh/ruff/pull/11485, so it make sense?  

Also, in the https://github.com/teemtee/tmt/pull/2842/commits/b4fd64ec1cdcce7079a123db0b31f619829f6922, if you hate the `elif`s, so I know if I should drop it or not :)

thrix commented on pyproject.toml:
```diff
@@ -345,6 +351,9 @@ lint.ignore = [
     ]
 
+[tool.ruff.lint.per-file-ignores]
+"tests/unit*" = ["S604", "S605", "S318", "S108"]  # Less strict security checks in tests/
```
@martinhoyer can we list this out nicely with comments?

martinhoyer replied:
Addressed in https://github.com/teemtee/tmt/pull/2842/commits/706c670a649614f3d616e80b39120af29202101f

thrix commented on tmt/cli.py:
```diff
@@ -2030,20 +2030,21 @@ def setup_completion(shell: str, install: bool) -> None:
     if install:
-        out.close()
+        with open(script, 'w') as f:
+            subprocess.run(command, stdout=f)
```
`shell=True` was dropped, what seems like a bug. I am all in dropping it, but then, we need to use `env` to pass the variable into the environment.

martinhoyer replied:
Thanks. Adding it back, as you mentioned, we would need env.
Do we want to disable the [S602](https://docs.astral.sh/ruff/rules/subprocess-popen-with-shell-equals-true/) completely?

martinhoyer replied:
Addressed in https://github.com/teemtee/tmt/pull/2842/commits/706c670a649614f3d616e80b39120af29202101f

Let me know if `noqa` would be preferred, so ruff checks for `shell=True` going forward

thrix replied:
I think in general, when possible, we should avoid using it, as it can open some holes in theory.
In this case, seems the only thing we need is adding another environment variable, in that case I would rather go with `env`, copy the process environment and add the one we need.

@WDYT @teemtee/tmt

happz replied:
Dropping `shell` does sound like a good idea. If all we need is to add an environment variable to the mix, then `env` is exactly the answer.

lukaszachy replied:
+1 to thrix' proposal.

S602 is good check to  have however in tmt we are expected to run as shell sometimes. But we need to be aware that it is the case and, maybe, over time find way to avoid doing so ;)

LecrisUT replied:
Had to look at the implications of `shell=True`, and from what I understand it is to make sure a single command is being run
```python
# insecure
some_insecure_flags="-la; do_some_weird_stuff"
subprocess.run(f"ls {some_insecure_flags}", shell=True)
# more secure
sanitized_flags=shlex.split(some_insecure_flags)
subprocess.run(["ls", *sanitized_flags])
```
But I think in the case for tmt, we generally want to enable the "insecure" approach in order to run a yaml text block, but probably this can be scoped specifically to running the test scripts parts. As for `env`, I am not sure, you mean the part about prefixing `f"{COMPLETE_VARIABLE}={shell}_source"`?

happz replied:
Yep, sometimes tmt needs to run commands with `shell=True`, but this does not seem to be the case. `f'{COMPLETE_VARIABLE}={shell}_source tmt'` - no redirection, no shell magic, no flow controls, just a command with environment spiked with a single extra variable. `Command('tmt').run(env={COMPLETE_VARIABLE: f'{shell}_source'})`, done :)

martinhoyer replied:
> `Command('tmt').run(env={COMPLETE_VARIABLE: f'{shell}_source'})`, done :)

oh, that's way better. Doing that.

happz replied:
Yep, I for one would go with `tmt.Command` (or `tmt.utils.Command`) as the way to run external commands in the tmt codebase.

happz replied:
LGTM.

thrix commented on tmt/steps/provision/__init__.py:
```diff
@@ -778,7 +778,8 @@ def _random_name(self, prefix: str = '', length: int = 16) -> str:
         name = prefix + ''.join(
-            random.choices(string.ascii_letters, k=min_random_part))
+            # S311: Use of pseudo-random generator is adequate for tmt run id
+            random.choices(string.ascii_letters, k=min_random_part))  # noqa: S311
```
@martinhoyer maybe we could just use the secrets module? The no noqa magic needed

Note that this will require `import secrets` at the start

```suggestion:-1
            ''.join(secrets.choice(string.ascii_letters) for _ in range(min_random_part))
```

thrix commented on tmt/steps/provision/__init__.py:
```diff
@@ -778,7 +778,8 @@ def _random_name(self, prefix: str = '', length: int = 16) -> str:
         min_random_part = max(5, length - len(prefix))
         name = prefix + ''.join(
-            random.choices(string.ascii_letters, k=min_random_part))
+            # S311: Use of pseudo-random generator is adequate for tmt run id
+            random.choices(string.ascii_letters, k=min_random_part))  # noqa: S311
```
@martinhoyer maybe we could just use the secrets module? The no noqa magic needed

Note that this will require `import secrets` at the start

```suggestion
            ''.join(secrets.choice(string.ascii_letters) for _ in range(min_random_part))
```

martinhoyer replied:
Thanks. Interestingly it even looks like it's not more expensive.

thrix commented on tmt/base.py:
```diff
@@ -2917,34 +2917,32 @@ def name_filter(nodes: Iterable[fmf.Tree]) -> list[fmf.Tree]:
                     keys=keys, sources=cmd_line_names)]
-
-        else:
+        elif not unique and names:
```
@martinhoyer something @happz is preparing a styling guide afaik, prefer a space before rather ...

thrix replied:
and the next space can be removed, as it was before

martinhoyer replied:
Thanks. Added the space back.

thrix commented on tmt/cli.py:
```diff
@@ -2034,7 +2034,7 @@ def setup_completion(shell: str, install: bool) -> None:
     if install:
         with open(script, 'w') as f:
-            subprocess.run(command, stdout=f)
+            subprocess.run(command, stdout=f, check=False)
```
why adding `check=False` here?

martinhoyer replied:
If I remember correctly, it was about explicitly setting it instead of relying on default value, but now when I remove it, nothing complains. So I guess it is indeed redundant.

thrix replied:
I would keep the original code myself, but no strong preference :)

martinhoyer replied:
> I would keep the original code myself, but no strong preference :)

Do you mean the Path.write_text I've changed this to in the last commit, or just removing the `check=False`?

martinhoyer commented on tmt/convert.py:
```diff
@@ -612,8 +611,8 @@ def target_content_build() -> list[str]:
 
-    log.debug('Common metadata:\n' + format_value(common_data))
-    log.debug('Individual metadata:\n' + format_value(individual_data))
+    log.debug(f'Common metadata:\n {format_value(common_data)}')
+    log.debug(f'Individual metadata:\n {format_value(individual_data)}')
```
Thanks! Addressed in https://github.com/teemtee/tmt/pull/2842/commits/3a62d23b1e7ba33003ea91077093034c6dc059dd

therazix replied:
@martinhoyer It looks like the spaces are still there. Was this change lost during the rebase or is this correct?

martinhoyer replied:
> @martinhoyer It looks like the spaces are still there. Was this change lost during the rebase or is this correct?

Damn, thanks a lot! Indeed looks like a botched rebase. Will address it when rebasing to main.

martinhoyer commented on tmt/export/nitrate.py:
```diff
@@ -264,8 +264,8 @@ def return_markdown_file() -> Optional[Path]:
     md_files = re.findall(reg_exp, files, re.M)
-    fail_message = "in the current working directory.\n" \
-                   "Manual steps couldn't be exported"
+    fail_message = ("in the current working directory.\n "
```
Thanks! Addressed in https://github.com/teemtee/tmt/pull/2842/commits/3a62d23b1e7ba33003ea91077093034c6dc059dd

thrix reviewed (approved):
LGTM

LecrisUT commented on tmt/cli.py:
```diff
@@ -2030,20 +2028,30 @@ def setup_completion(shell: str, install: bool) -> None:
 
     if install:
-        out.close()
+        Path(script).write_text(str(completions))
```
```suggestion
        assert completions is not None
        Path(script).write_text(completions)
```

How about doing type narrowing (probably a bit higher up) instead of the `str()`

martinhoyer replied:
damn, I left it there by accident. Thanks! You guys are so good at reviewing...   
The assert is not needed, as there is a check higher up.

LecrisUT commented on tmt/utils.py:
```diff
@@ -3920,7 +3919,7 @@ def format_value(
 def format(
         key: str,
-        value: Union[None, int, float, bool, str, list[Any], dict[Any, Any]] = None,
+        value: Union[None, float, bool, str, list[Any], dict[Any, Any]] = None,
```
Why is this change? Is it because `int` can be treated like `float`?

martinhoyer replied:
Yes, this rule to be specific: https://docs.astral.sh/ruff/rules/redundant-numeric-union/

happz replied:
But... but... why stopping there, `True` and `False` are also integers :)

happz replied:
> Lol, writing C code? :D

```
$ python
Python 3.11.9 (main, Apr 17 2024, 00:00:00) [GCC 13.2.1 20240316 (Red Hat 13.2.1-7)] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> type(True)
<class 'bool'>
>>> isinstance(True, int)
True
>>> 
```

> I think if it's to indicate that something is JSON-like stuff, it would be good to be more explicit about that, but I think list[Any] and dict[Any, Any] indicate that it isn't?

Not necessarily JSON, just values. I didn't want to use `Any`, as `format_value()` certainly cannot format everything, just primitives and dicts and lists. It would be better if even these `list` and `dict` wouldn't use `Any`, but also `Union[None, float, bool, str]`, but I guess that's for another patch

LecrisUT commented on tmt/cli.py:
```diff
@@ -2030,20 +2028,30 @@ def setup_completion(shell: str, install: bool) -> None:
-                   shell=True, stdout=out)
+    env_var = {COMPLETE_VARIABLE: f'{shell}_source'}
+
+    logger = context.obj.logger
+
+    completions = Command('tmt').run(env=tmt.utils.Environment.from_dict(env_var),
+                                     cwd=None,
+                                     logger=context.obj.logger
+                                     ).stdout
```
An issue for a future PR, do we really need to call `tmt` recursively? Feels weird to have a nested logger being created here.

happz replied:
No comments from my side.

LecrisUT commented on tmt/log.py:
```diff
@@ -797,13 +796,13 @@ def debug(
             )
 
-    def warn(
+    def warning(
```
Just to be safe, can you create a `def warn` and add `DeprecationWarning` in it? I think I was using it at some point when writing external plugins.

martinhoyer replied:
Thanks. Added in https://github.com/teemtee/tmt/pull/2842/commits/4524b65129be20c3dd42ac48b9333fe0e8889ef9

LecrisUT commented on tmt/steps/__init__.py:
```diff
@@ -1016,7 +1014,7 @@ class (single class or tuple of classes).
 
         if classes is None:
-            _classes: tuple[Union[type[Phase], type[PhaseT]], ...] = (Phase,)
+            _classes: tuple[type[Union[Phase, PhaseT]], ...] = (Phase,)
```
:eyes: didn't know this is transitive. Makes sense in retrospect.

LecrisUT commented on tmt/steps/provision/__init__.py:
```diff
@@ -778,7 +778,7 @@ def _random_name(self, prefix: str = '', length: int = 16) -> str:
         min_random_part = max(5, length - len(prefix))
         name = prefix + ''.join(
-            random.choices(string.ascii_letters, k=min_random_part))
+            secrets.choice(string.ascii_letters) for _ in range(min_random_part))
```
Feels like there can be a simpler approach here.

martinhoyer replied:
felt to me too, but this is basically "best practices" from python secrets documentation and was suggested by @thrix (and I haven't found anything better fwiw)

LecrisUT commented on tmt/steps/discover/fmf.py:
```diff
@@ -278,11 +278,11 @@ def is_in_standalone_mode(self) -> bool:
         return super().is_in_standalone_mode
 
-    def get_git_root(self, dir: Path) -> Path:
+    def get_git_root(self, directory: Path) -> Path:
```
Just FTR, since I commented about adding `def warn`, this should be fine.

LecrisUT reviewed (approved):
LGTM, thanks for working on this @martinhoyer. Only 1 real comment for this PR about `def warn` other than that, just comments for the future.

happz commented:
Apparently unrelated errors, we'll look into improving `virtual` tests eventually. Merging.
